keps/sig-network/2104-kube-proxy-lib/README.md

```diff
@@ -0,0 +1,280 @@
+- Exemplify the use of such a repository in a mock "backend" which uses the repository to process and respond to changes in the Kubernetes networking state.
+- Define a policy around versioning and releasing of "kube-proxy-lib".
+- Create a CI system that runs in test-grid which tests kube-proxy-lib compatibility with the latest Kubernetes releases continuously, and which can be used to inform API changes for regressions that break "kube-proxy-lib".
+- Enable the eventual *replacement* of the k/k/pkg/proxy serviceChangeTracker and endpointsChangeTracker related caching structures inside of in-tree kube proxy with this generic library.
+
+### Non-Goals
+
+- Rewrite or decouple the core , in-tree linux Kubernetes kube-proxy implementations, which are relied on by too many users to be tolerant to major changes.
+- Force a new architecture for the standard kube-proxy on to naive users.
+
+## Proposal
+
+We propose to build a kubernetes-sigs/kube-proxy-lib repository.  This repository will be vendored by people wanting to build a new Kube proxy, and provide them with:
+- A vendorable golang library that defines a few interfaces that can be easily implemented as a new kube proxy, which respond to endpoint slices and services changes.
+- Documentation on how to build a kube proxy, based on https://github.com/kubernetes-sigs/kpng/blob/master/doc/service-proxy.md and other similar documents.
+- integration test tooling, similar to the KPNG project's, which allows users to locally implement network routing logic in a small golang program which is not
+directly connected to the Kubernetes API server, for local, iterative development of Kubernetes network proxy tooling.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a networking technology startup I want to make my own kube-proxy implementation but don't want to maintain the logic of talking to the APIServer, caching its data, or calculating an abbreviated/proxy-focused representation of the Kubernetes networking state space.  I'd like a wholesale framework I can simply plug my logic into. 
+
+#### Story 2
+
+As a Kubernetes maintainer, I don't want to have to understand the internals of a networking backend in order to simulate or write core updates to the logic of the kube-proxy locally.
+
+### Notes/Constraints/Caveats (Optional)
+
+- sending the full-state could be resource consuming on big clusters, but it should still be O(1) to
+  the actual kernel definitions (the complexity of what the node has to handle cannot be reduced
+  without losing functionality or correctness).
+
+### Risks and Mitigations
+
+No risks, because we arent removing the in tree proxy as part of this repo, but rather, proposing a library for kube proxy extenders
+to use optionally.  There may be risks, eventually, if we write another KEP to *replatform* sig-networks k/k/pkg/proxy implementations
+on top of this library, but that would be in a separate KEP. 
```

Comment:
Fair point but, in the interest of "building something that is uncontraversial as a first step", How strongly do you feel about this ?  If i made this argument:


- "sig network needs b"
- "b needs a"
- thus were going to make "sig-network/a" as a stepping stone to us being able to build "sig-network/b" 
- by the way "sig-network/b" is somethin lots of people want, so even if we get hit by a bus after making "b", and we never get around to "a", we still have a useful intermediate deliverable.... 

Is that digestible ? or does it sit funny w/ you ?

Reply:
What I'm saying is that I don't think the library can be definitely "uncontroversial" and "a useful intermediate deliverable" unless we _know_ that it is suited to replacing `pkg/proxy`.

I'm not saying we have to deliver the library and the updated iptables proxy in the same release, I'm saying that rather than doing:
1. come up with a plan for the library
2. implement the library
3. come up with a plan to replace `pkg/proxy` with the new library
4. replace `pkg/proxy` with the new library

we need to do
1. come up with a plan for the library
2. come up with a plan to replace `pkg/proxy` with the new library
3. implement the library
4. replace `pkg/proxy` with the new library

Reply:
gotcha, ok updated as a risk (and updated the goals to reflect)